---
title: "Heatmap demo"
author: "Mira Sohn"
output:
    html_document:
        keep_md: yes
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r, include=FALSE}
knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE)
```


This workflow is designed to practice data visualization using heatmap based on this tutorial:     https://jcoliver.github.io/learn-r/008-ggplot-dendrograms-and-heatmaps.html    



## Load packages

Links to package manuals:     
    - [ggplot2](https://ggplot2.tidyverse.org/)       
    - [tidyverse](https://www.tidyverse.org/)      
    - [viridis](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html)      
    - [ggdendro](https://cran.r-project.org/web/packages/ggdendro/vignettes/ggdendro.html)      
    - [DT](https://rstudio.github.io/DT/)      


```{r packages}
library(ggplot2)   # plotting
library(tidyverse) # data manipulation
library(viridis)   # color palette
library(ggdendro)  # dendrogram
library(DT)        # interative table presentation

set.seed(5678)
```

## Set user variables

User variables include factor levels, color palette, font style/size, and so on. They can be moved to a separate configuration or wrapper file if you prefer to have them externally. I like to minimize hard-coded variables for fine-tuning the plots in the future.

```{r user_variable}


# Set user variables for heatmap
heatmap.font <- 12
heatmap.palette <- colorRampPalette(viridis::magma(7))(100)
user.font <- "Helvetica"
my.genotypes <- c("WT", "KO")



# Define colors for legend
anno.color <- list(Genotype=c(WT="#1b94a5", KO="#2dbace"))

# Ratio between ggplot and dendrogram
gg.dendro.ratio <- c(0.8, 0.2)


```


## Import input

Assume there are 20 top genes found in RNA-seq analysis. Imported data is log-transformed normalized read count table. It's been pre-processed to avoid additional data cleaning steps prior to plotting. 

```{r import_data, results='asis'}

# Import input data
mtx <- read.csv("test.csv") %>%
  column_to_rownames("X")

# Print the input count matrix
DT::datatable(mtx)

# Build input read count matrix:
# psuedocount + 1 and log-transformation
# conversion is needed for pheatmap() and dist() which only take matrix obj
mtx <- as.matrix(mtx)
```



## Creat heatmap using pheatmap package

pheatmap is one of the most well-known package providing heatmap with dendrogram from hierarchical clustering of genes and samples. 


```{r pheatmap, results='asis', fig.height=8, fig.width=8}

# Use pheatmap() from pheatmap package

# Build input metadata
heat.meta <- data.frame(Sample=colnames(mtx)) %>%
    separate(col=Sample, c("Genotype", "Celltype", "Replicate"), sep="_", remove=FALSE) %>%
    mutate(Replicate=as.factor(Replicate)) 

rownames(heat.meta) <- heat.meta$Sample

# Update metadata by filtering out any columns having a single factor
col.include <- apply(heat.meta, 2, function(x) { length(unique(x)) }) != 1
heat.meta <- heat.meta[, col.include]

heat.meta$Genotype <- factor(heat.meta$Genotype, levels=c("WT", "KO"))
heatmap1 <-
    pheatmap::pheatmap(mtx,
     show_colnames=TRUE,
     show_rownames=TRUE,
     annotation_col=heat.meta[,colnames(heat.meta) != "Sample"],
     cutree_cols=2,
     border_color=NA,
     color = heatmap.palette,
     annotation_colors=anno.color,
     annotation_names_col=F,
     fontsize=heatmap.font)

heatmap1


```

## Create heatmap using ggplot2 and ggdendro packages

Built-in font styles are found here: http://www.cookbook-r.com/Graphs/Fonts/      
If you'd like to use diverse custom font styles, use font packages such as [extrafont](https://cran.r-project.org/web/packages/extrafont/).


```{r ggplot2, results='asis', fig.height=8, width=8}


# Use ggplot() from ggplot2 package & ggdendero


# Create dendrogram
dendro <- as.dendrogram(hclust(dist(mtx)))
dendro.plot <- ggdendrogram(dendro,
                            rotate=TRUE) + 
    theme(axis.text.y=element_blank(),
          axis.text.x=element_blank())

# Extract rows (genes) from dendrogram
dendro.order <- order.dendrogram(dendro)



# Clean input data frame
counts.df <- as.data.frame(mtx) %>%
    rownames_to_column(var="Gene") %>%
    gather("Sample", "Count", -Gene) %>%
    mutate(Sample=factor(Sample, levels=colnames(mtx))) %>%
    separate(Sample, c("Genotype", "Celltype", "Replicate"), remove=F) %>%
    mutate(Sample=str_replace_all(Sample, "_", " "),
           Genotype=factor(Genotype, levels=my.genotypes)) %>%
    dplyr::rename(`Count(log)` = Count)



# Set levels of counts.df$Gene to genes ordered in dendrogram
counts.df$Gene <- factor(counts.df$Gene, levels=counts.df$Gene[dendro.order])


# Create heatmap
heatmap2 <-
    ggplot(counts.df,
       aes(x=Replicate, y=Gene, fill=`Count(log)`)) +
geom_tile() +
scale_fill_viridis() +
theme_bw() +
theme(text=element_text(family=user.font, size=heatmap.font),   # family is used to change font style
      legend.position="top",
      legend.title=element_text(size=heatmap.font),
      strip.placement="outside",
      strip.text.x=element_text(size=heatmap.font)) +
facet_grid(~ Genotype, scales="free_x", space="free_x")


# Print the plots
grid::grid.newpage()
print(heatmap2, vp=grid::viewport(x=0.4, y=0.5, width=gg.dendro.ratio[1], height=1))
print(dendro.plot, vp=grid::viewport(x=0.9, y=0.448, width=gg.dendro.ratio[2], height=0.871))


```

Aligning the two graphs are the hardest part of this workflow. Consult the tutorial introduced above.


## Save the individual output plots

Individual output plots from `pheatmap()`, `ggplot()`, and `ggdendrogram()` are saved using `ggsave()` from the ggplot2 package. Here, I created `heatmap1` from `pheatmap()`, `heatmap2` from `ggplot()`, and `dendro.plot` from `ggdendrogram()`. 

```{r save_individual, eval=FALSE}

# Save individual plots
ggsave("heatmap1.pdf", heatmap1, device="pdf")  # returns pdf format in my current directory
ggsave("heatmap2.pdf", heatmap2, device="pdf")
ggsave("dendro.pdf", dendro.plot, device="pdf", width=2)  # width=2 is set to limit width of dendrogram

```

## Save the ggplot-ggdendro aligned plot

The aligned plots are saved using `pdf()` and `dev.off()`.

```{r save_combined, eval=FALSE}

# Save aligned plots
pdf("aligned.heatmap.pdf"), fonts=user.font, width=8, height=8)
grid::grid.newpage()
print(heatmap2, vp=grid::viewport(x=0.4, y=0.5, width=gg.dendro.ratio[1], height=1))
print(dendro.plot, vp=grid::viewport(x=0.9, y=0.448, width=gg.dendro.ratio[2], height=0.871))
dev.off()
```


## Session info
For reproducibility purposes, here is the output of `sessionInfo()` showing the
versions of all packages used here.

```{r sessioninfo, collapse=FALSE}
sessionInfo()
```


